dev/UndockedRegFreeWinRT/typeresolution.cpp

```diff
@@ -428,40 +428,69 @@ namespace UndockedRegFreeWinRT
-        // which will allow type resolution against adjacent WinMDs.
-        if (hr == HRESULT_FROM_WIN32(APPMODEL_ERROR_NO_PACKAGE))
+        // Walk the package graph looking for the requested metadata
+        const uint32_t filter{ PACKAGE_FILTER_HEAD | PACKAGE_FILTER_DIRECT | PACKAGE_FILTER_IS_IN_RELATED_SET | PACKAGE_FILTER_DYNAMIC | PACKAGE_FILTER_STATIC };
+        uint32_t bufferLength{};
+        uint32_t packagesCount{};
+        bool processHasStaticPackageGraph{};
+        HRESULT hr{ HRESULT_FROM_WIN32(GetCurrentPackageInfo(filter, &bufferLength, nullptr, &packagesCount)) };
+        if (hr == HRESULT_FROM_WIN32(ERROR_INSUFFICIENT_BUFFER))
         {
-            PCWSTR exeDir = nullptr;  // Never freed; owned by process global.
-            RETURN_IF_FAILED(GetProcessExeDir(&exeDir));
+            // The process has a package graph. Walk it looking for the requested metadata
+            auto buffer{ wil::make_unique_nothrow<BYTE[]>(bufferLength) };
+            RETURN_IF_NULL_ALLOC(buffer);
+            RETURN_IF_WIN32_ERROR(GetCurrentPackageInfo(filter, &bufferLength, buffer.get(), &packagesCount));
+            const auto packageInfos{ reinterpret_cast<PACKAGE_INFO*>(buffer.get()) };
+            for (uint32_t index=0; index < packagesCount; ++index)
+            {
+                const auto& packageInfo{ packageInfos[index] };
+                HRESULT hrFindType{ FindTypeInDirectoryWithNormalization(pMetaDataDispenser, pszFullName,
+                                packageInfo.path, phstrMetaDataFilePath, ppMetaDataImport, pmdTypeDef) };
+                if (SUCCEEDED(hrFindType))
+                {
+                    return hrFindType;
+                }
+
+                // Keep track if we find any Static entries in the package graph
+                if (WI_IsFlagSet(packageInfo.flags, PACKAGE_PROPERTY_STATIC))
+                {
+                    processHasStaticPackageGraph = true;
+                }
+            }
+        }
+
+        // Not found in the package graph.
 
+        // Unpackaged apps may have metadata next to their executable.
+        //
+        // Packaged apps have metadata in their root directory which is
+        // in the package graph and thus already checked.
+        //
+        // NOTE: Only packaged apps have Static entries in their package graph.
+        //       Unpackaged apps have no Static entries in their package graph.
+        if (!processHasStaticPackageGraph)
+        {
+            PCWSTR exeDir{};  // Never freed; owned by process global.
+            RETURN_IF_FAILED(GetProcessExeDir(&exeDir));
             hr = FindTypeInDirectoryWithNormalization(
                 pMetaDataDispenser,
                 pszFullName,
                 exeDir,
                 phstrMetaDataFilePath,
                 ppMetaDataImport,
                 pmdTypeDef);
-
             if (hr == RO_E_METADATA_NAME_NOT_FOUND)
```

Comment:
RETURN_HR_IF would make this more concise and add WIL logging.

Reply:
Except we don't want to log a WIL message in this case. We could invert the logic

```
// ...comment block...
RETURN_HR_IF(hr, hr != RO_E_METADATA_NAME_NOT_FOUND);
return HRESULT_FROM_WIN32(APPMODEL_ERROR_NO_PACKAGE);
```

but reading double-negatives is a bit more cryptic so not a helpful change in this case